src/vegasflow/vflowplus.py

```diff
@@ -0,0 +1,225 @@
+#!/usr/env python
+"""
+    Implementation of vegas+ algorithm:
+   
+    adaptive importance sampling + adaptive stratified sampling
+         
+"""
+from itertools import product
+import tensorflow as tf
+import numpy as np
+from vegasflow.configflow import DTYPE, DTYPEINT, fone, fzero, float_me, ione, int_me, BINS_MAX
+from vegasflow.monte_carlo import wrapper
+from vegasflow.vflow import VegasFlow
+from vegasflow.utils import consume_array_into_indices
+
+import logging
+
+logger = logging.getLogger(__name__)
+
+BETA = 0.75
+FBINS = float_me(BINS_MAX)
+
+@tf.function(input_signature=3 * [tf.TensorSpec(shape=[None, None], dtype=DTYPE)])
+def _compute_x(x_ini, xn, xdelta):
+    """ Helper function for ``generate_samples_in_hypercubes`` """
+    aux_rand = xn - tf.math.floor(xn)
+    return x_ini + xdelta * aux_rand
+
+# same as vegasflow generate_random_array
+@tf.function(input_signature=2*[tf.TensorSpec(shape=[None, None], dtype=DTYPE)])
+def _digest(xn, divisions):
+    ind_i = tf.cast(xn, DTYPEINT)
+    # Get the value of the left and right sides of the bins
+    ind_f = ind_i + ione
+    x_ini = tf.gather(divisions, ind_i, batch_dims=1)
+    x_fin = tf.gather(divisions, ind_f, batch_dims=1)
+    # Compute the width of the bins
+    xdelta = x_fin - x_ini
+    return ind_i, x_ini, xdelta
+
+
+@tf.function(
+    input_signature=[
+        tf.TensorSpec(shape=[None, None], dtype=DTYPE),
+        tf.TensorSpec(shape=[], dtype=DTYPEINT),
+        tf.TensorSpec(shape=[None], dtype=DTYPEINT),
+        tf.TensorSpec(shape=[None, None], dtype=DTYPE),
+        tf.TensorSpec(shape=[None, None], dtype=DTYPE),
+    ]
+)
+def generate_samples_in_hypercubes(rnds, n_strat, n_ev, hypercubes, divisions):
+    """Receives an array of random numbers 0 and 1 and
+    distribute them in each hypercube according to the
+    number of sample in each hypercube specified by n_ev
+
+    Parameters
+    ----------
+        `rnds`: tensor of random number betweenn 0 and 1
+        `n_strat`: tensor with number of stratifications in each dimension
+        `n_ev`: tensor containing number of sample per each hypercube
+        `hypercubes`: tensor containing all different hypercube
+        `divisions`: vegas grid
+
+    Returns
+    -------
+        `x` : random numbers collocated in hypercubes
+        `w` : weight of each event
+        div_index: division index in which each (n_dim) set of random numbers fall
+        `segm` : segmentantion for later computations
+    """
+
+    points = tf.repeat(hypercubes, n_ev, axis=0)
+    n_evs = float_me(tf.repeat(n_ev, n_ev))
+    xn = tf.transpose((points + tf.transpose(rnds)) * FBINS / float_me(n_strat))
```

Comment:
The crash happens here, where the large tensors are generated when calling `tf.repeat`.
This is due to the fact that after the first call of `redistribute_samples`, using large number of events, there may be some hypercubes with a large number of samples, around 1e4 samples per hypercube.
In Lepage's implementation there is a parameter `max_neval_hcube` which is the maximum number of integrand evaluations per hypercube, which is set to `1e7` that would be too large to solve this issue.
Should I implement a smaller `max_neval_hcube` around `1e3` to avoid these large tensors or maybe there is a different way of generating samples in hypercubes without using `tf.repeat`?

Reply:
Which of the two repeats is giving the problem? I think there should be an alternative method to do this since at the end of the day you won't have more points than evaluations.

Reply:
The first one, in fact the shape `[65536,51749,8]` corresponds to `65536` hypercubes, `8` is the dimensions and `51749` are the evaluations in one particular hypercube.

Reply:
There is a bug somewhere because if I repeat hypercubes, according to n_ev (which should have the same size) along the axis 0 I should get a rank 2 tensor and that's a rank 3 (because what you are doing is to use each of the `n_ev` to repeat each of the `hypercubes`.

But the thing is, I just tried in my laptop to do that repeat with 1e6 events per hypercube and it worked (I tried with 2 dimensions and I get a 2250000000x2 tensor).

Reply:
> But the thing is, I just tried in my laptop to do that repeat with 1e6 events per hypercube and it worked (I tried with 2 dimensions and I get a 2250000000x2 tensor).

In lower dimensions works fine in my laptop too.

> Edit: trying to reproduce your error, could which example are you running? Which version of tensorflow?

Tensorflow 2.4.1
Python 3.7.3
The example is retracing.py using 1e6 samples and dim 8.

Reply:
I can reproduce the error. Strange though because by size the array I was creating is bigger than the one that is failing.

In any case, the strategy of repeating hypercubes might not be the best because in any case the memory gets cluttered. Since this is only used for `xn` I wonder whether one could play around with gather or scatter to get the answer without the intermediate array... I'm busy now but if you don't find a solution I'll have a look tomorrow. Worst case scenario we'll have to write our own operator (since the problem is internal to repeat).

Reply:
I understand. I'll see if there is another method to play around it.
Thanks for helping!

Reply:
Rethinking about this, creating a custom operator to sum rnds and the hypercubes according to n_ev might be the best thing to do since the content of n_ev changes iteration to iteration... I'm not entirely Tensorflow has a method to do that efficiently (since TF is geared towards ML and you wouldn't expect the architecture to change iteration to iteration).

Reply:
Just to be clear, by creating a custom operator you mean something like creating a function in python without using the problematic TF primitives such as tf.repeat or writing the operator directly C++?

Reply:
Directly in C++ and Cuda (for CPU and GPU).
But if it can be done with TF primitives, even if it is not optimal, it's a good starting point (and the custom operators can be left for the future).

Reply:
Ok, thanks!
I'll start by using TF primitives for now.

Reply:
Btw, if you put a very big number of dimensions you don't get the error (because the number of hypercubes goes down to 1). I guess a limit is needed (such as the 1e6 from Lepage) maybe it is better to be more restrictive for `n_strat` (since I would guess the adaptation part would not be very efficient for ~10 dimensions).

Reply:
Yes, sure. In Lepage's code there is a limit for the max number of hypercubes `max_nhcube = 1e9` ; which does not solve the problem, since is too big,  but it can be useful.

Reply:
I try to fix the problem by using `tf.repeat` only to generate the indices and the applying `tf.gather` to obtain the variable `points`. In this way it doesn't generate the big rank 3 tensor which was causing the crash.
```    
    # memory problem: tf.repeat generates rank 3 tensor too big
    #points = tf.repeat(hypercubes, n_ev, axis=0)
    #n_evs = float_me(tf.repeat(n_ev, n_ev))


    #alternative apply repeat only to indices:
    indices = tf.repeat(tf.range(tf.shape(hypercubes,out_type=tf.int64)[0]),n_ev)
    points = float_me(tf.gather(hypercubes,indices))
    n_evs = float_me(tf.gather(n_ev,indices))
```
There are still problems regarding `tf.repeat`, sometimes it works and other times I get one of the following errors:
<details>
<summary>Error1</summary>
<pre>2021-04-10 08:35:34.498445: W tensorflow/core/framework/op_kernel.cc:1763] OP_REQUIRES failed at reshape_op.h:57 : Invalid argument: Size 0 must be non-negative, not -1031536640
Traceback (most recent call last):
  File "examples/retracing.py", line 43, in <module>
    vegas_instance.run_integration(n_iter)
  File "/Users/andreapasquale/Desktop/vegasflow/src/vegasflow/monte_carlo.py", line 428, in run_integration
    res, error = self._run_iteration()
  File "/Users/andreapasquale/Desktop/vegasflow/src/vegasflow/vflow.py", line 366, in _run_iteration
    res, sigma = self.iteration_content()
  File "/Users/andreapasquale/Desktop/vegasflow/src/vegasflow/vflowplus.py", line 208, in _iteration_content
    ress, arr_var, arr_res2 = self.run_event(n_ev=self.n_ev)
  File "/Users/andreapasquale/Desktop/vegasflow/src/vegasflow/vflowplus.py", line 226, in run_event
    return super().run_event(tensorize_events=self.adaptive, **kwargs)
  File "/Users/andreapasquale/Desktop/vegasflow/src/vegasflow/monte_carlo.py", line 329, in run_event
    res = self.device_run(ncalls, sent_pc=pc, **kwargs)
  File "/Users/andreapasquale/Desktop/vegasflow/src/vegasflow/monte_carlo.py", line 248, in device_run
    result = self.event(ncalls=ncalls, **kwargs)
  File "/Users/andreapasquale/anaconda3/envs/python38/lib/python3.8/site-packages/tensorflow/python/eager/def_function.py", line 824, in __call__
    return self._python_function(*args, **kwds)
  File "/Users/andreapasquale/Desktop/vegasflow/src/vegasflow/monte_carlo.py", line 381, in run_event
    return self._run_event(tf_integrand, **kwargs)
  File "/Users/andreapasquale/Desktop/vegasflow/src/vegasflow/vflowplus.py", line 177, in _run_event
    x, ind, w, segm = generate_samples_in_hypercubes(
  File "/Users/andreapasquale/anaconda3/envs/python38/lib/python3.8/site-packages/tensorflow/python/eager/def_function.py", line 824, in __call__
    return self._python_function(*args, **kwds)
  File "/Users/andreapasquale/Desktop/vegasflow/src/vegasflow/vflowplus.py", line 77, in generate_samples_in_hypercubes
    indices = tf.repeat(tf.range(tf.shape(hypercubes,out_type=tf.int64)[0]),n_ev)
  File "/Users/andreapasquale/anaconda3/envs/python38/lib/python3.8/site-packages/tensorflow/python/util/dispatch.py", line 201, in wrapper
    return target(*args, **kwargs)
  File "/Users/andreapasquale/anaconda3/envs/python38/lib/python3.8/site-packages/tensorflow/python/ops/array_ops.py", line 6336, in repeat
    return repeat_with_axis(input, repeats, axis, name)
  File "/Users/andreapasquale/anaconda3/envs/python38/lib/python3.8/site-packages/tensorflow/python/ops/array_ops.py", line 6236, in repeat_with_axis
    masked = boolean_mask(tiled, mask)
  File "/Users/andreapasquale/anaconda3/envs/python38/lib/python3.8/site-packages/tensorflow/python/util/dispatch.py", line 201, in wrapper
    return target(*args, **kwargs)
  File "/Users/andreapasquale/anaconda3/envs/python38/lib/python3.8/site-packages/tensorflow/python/ops/array_ops.py", line 1754, in boolean_mask
    tensor = reshape(
  File "/Users/andreapasquale/anaconda3/envs/python38/lib/python3.8/site-packages/tensorflow/python/util/dispatch.py", line 201, in wrapper
    return target(*args, **kwargs)
  File "/Users/andreapasquale/anaconda3/envs/python38/lib/python3.8/site-packages/tensorflow/python/ops/array_ops.py", line 195, in reshape
    result = gen_array_ops.reshape(tensor, shape, name)
  File "/Users/andreapasquale/anaconda3/envs/python38/lib/python3.8/site-packages/tensorflow/python/ops/gen_array_ops.py", line 8368, in reshape
    _ops.raise_from_not_ok_status(e, name)
  File "/Users/andreapasquale/anaconda3/envs/python38/lib/python3.8/site-packages/tensorflow/python/framework/ops.py", line 6862, in raise_from_not_ok_status
    six.raise_from(core._status_to_exception(e.code, message), None)
  File "<string>", line 3, in raise_from
tensorflow.python.framework.errors_impl.InvalidArgumentError: Size 0 must be non-negative, not -1031536640 [Op:Reshape]</pre>
</details>

<details>
<summary>Error2</summary>
<pre>Traceback (most recent call last):e: 100.0 %
  File "examples/retracing.py", line 43, in <module>
    vegas_instance.run_integration(n_iter)
  File "/Users/andreapasquale/Desktop/vegasflow/src/vegasflow/monte_carlo.py", line 428, in run_integration
    res, error = self._run_iteration()
  File "/Users/andreapasquale/Desktop/vegasflow/src/vegasflow/vflow.py", line 366, in _run_iteration
    res, sigma = self.iteration_content()
  File "/Users/andreapasquale/Desktop/vegasflow/src/vegasflow/vflowplus.py", line 206, in _iteration_content
    ress, arr_var, arr_res2 = self.run_event(n_ev=self.n_ev)
  File "/Users/andreapasquale/Desktop/vegasflow/src/vegasflow/vflowplus.py", line 224, in run_event
    return super().run_event(tensorize_events=self.adaptive, **kwargs)
  File "/Users/andreapasquale/Desktop/vegasflow/src/vegasflow/monte_carlo.py", line 329, in run_event
    res = self.device_run(ncalls, sent_pc=pc, **kwargs)
  File "/Users/andreapasquale/Desktop/vegasflow/src/vegasflow/monte_carlo.py", line 248, in device_run
    result = self.event(ncalls=ncalls, **kwargs)
  File "/Users/andreapasquale/anaconda3/envs/python38/lib/python3.8/site-packages/tensorflow/python/eager/def_function.py", line 824, in __call__
    return self._python_function(*args, **kwds)
  File "/Users/andreapasquale/Desktop/vegasflow/src/vegasflow/monte_carlo.py", line 381, in run_event
    return self._run_event(tf_integrand, **kwargs)
  File "/Users/andreapasquale/Desktop/vegasflow/src/vegasflow/vflowplus.py", line 175, in _run_event
    x, ind, w, segm = generate_samples_in_hypercubes(
  File "/Users/andreapasquale/anaconda3/envs/python38/lib/python3.8/site-packages/tensorflow/python/eager/def_function.py", line 824, in __call__
    return self._python_function(*args, **kwds)
  File "/Users/andreapasquale/Desktop/vegasflow/src/vegasflow/vflowplus.py", line 76, in generate_samples_in_hypercubes
    indices = tf.repeat(tf.range(tf.shape(hypercubes,out_type=tf.int64)[0]),n_ev)
  File "/Users/andreapasquale/anaconda3/envs/python38/lib/python3.8/site-packages/tensorflow/python/util/dispatch.py", line 201, in wrapper
    return target(*args, **kwargs)
  File "/Users/andreapasquale/anaconda3/envs/python38/lib/python3.8/site-packages/tensorflow/python/ops/array_ops.py", line 6336, in repeat
    return repeat_with_axis(input, repeats, axis, name)
  File "/Users/andreapasquale/anaconda3/envs/python38/lib/python3.8/site-packages/tensorflow/python/ops/array_ops.py", line 6236, in repeat_with_axis
    masked = boolean_mask(tiled, mask)
  File "/Users/andreapasquale/anaconda3/envs/python38/lib/python3.8/site-packages/tensorflow/python/util/dispatch.py", line 201, in wrapper
    return target(*args, **kwargs)
  File "/Users/andreapasquale/anaconda3/envs/python38/lib/python3.8/site-packages/tensorflow/python/ops/array_ops.py", line 1754, in boolean_mask
    tensor = reshape(
  File "/Users/andreapasquale/anaconda3/envs/python38/lib/python3.8/site-packages/tensorflow/python/util/dispatch.py", line 201, in wrapper
    return target(*args, **kwargs)
  File "/Users/andreapasquale/anaconda3/envs/python38/lib/python3.8/site-packages/tensorflow/python/ops/array_ops.py", line 195, in reshape
    result = gen_array_ops.reshape(tensor, shape, name)
  File "/Users/andreapasquale/anaconda3/envs/python38/lib/python3.8/site-packages/tensorflow/python/ops/gen_array_ops.py", line 8368, in reshape
    _ops.raise_from_not_ok_status(e, name)
  File "/Users/andreapasquale/anaconda3/envs/python38/lib/python3.8/site-packages/tensorflow/python/framework/ops.py", line 6862, in raise_from_not_ok_status
    six.raise_from(core._status_to_exception(e.code, message), None)
  File "<string>", line 3, in raise_from
tensorflow.python.framework.errors_impl.InvalidArgumentError: Input to reshape is a tensor with 4992729088 values, but the requested shape has 697761792 [Op:Reshape]</pre>
</details>

Maybe a possible solution could be to generate the `indices` variabile without using `tf.repeat`.

Reply:
What is the dtype of indices?

Reply:
I converted both `hypercubes` and `n_ev` to `tf.int64`.
And `indices` is also a `tf.int64` tensor.
I can commit this change if you want to test this implementation.

Reply:
Did you check the dtype delivered by tf.range?

Reply:
Yes, 
`tf.range(tf.shape(hypercubes,out_type=tf.int64)[0])` has dtype `tf.int64`
while for example
`tf.range(tf.shape(hypercubes)[0])` has dtype `tf.int32`

Reply:
Ok, after going a bit on circles for a while I'm convinced you cannot get better than the indices solution you wrote without going to custom operators. After talking with Stefano, I've been convinced this is a bad idea and that putting limits is better:

i.e:

```python
    indices = tf.repeat(tf.range(tf.shape(hypercubes,out_type=tf.int64)[0]),n_ev)
    points = float_me(tf.gather(hypercubes,indices))
    n_evs = float_me(tf.gather(n_ev,indices))
```

Maybe to avoid these errors would be to generate the indices outside this function, given that the number of hypercubes is constant from iteration to iteration you can have an integer variable saved which is the number of hypercubes and then in `redistribute_samples` do

```python
indices = tf.concat([tf.repeat(i, self.n_ev[i]) for i in range(self.n_hypercubes)]
```
(effectively this loop should be compiled out by tensorflow so it doesn't matter).

Doing it outside means you can have more fine-grained control on how big you allow this tensor to be.

However, performance-wise I don't think this is a good solution. Even if it works in CPU, the moment you got to the GPU I'd expect it to be terrible. And if you need to go up to `int64` I very much doubt the adaptation part would make a difference.

So in summary, I think that, unless the numeric performance is much better for a large number of dimensions, it would be best to put a stringent limit on the number of hypercubes (or points in hypercubes, depending which one works best).
If the numeric performance is much better such that having extremely large number of hypercubes is required, then we can work on the custom operator.

Reply:
Ok, I understand.
What remains to be done is to test the performance:
1. putting some kind of limit (number of hypercubes or samples in hypercubes) 
2. putting `indices` outside the `generate_samples_in_hypercubes` function

About the second one, I should pass `indices` as an argument to `generate_samples_in_hypercubes`, correct?
And to keep track of `indices` I guess I can add a `self.indices` attribute or it can create problems?

Reply:
I think right now what would be interesting is to see the performance on you can reach the same accuracy with less events, but don't worry about the cpu/gpu-performance. 

I'm starting to think that `tf.repeat` is broken. Or at least broken enough that it is not worth it to break the rest just so we can have `tf.repeat` in.
Internally at some point it gets above the limit for int32 even though the inputs were not big enough for that.

Anyway, I guess it's better to just set `run_eager(True)` and do something simple. I was able to integrate with:

```python
    indices = tf.concat([tf.repeat(i, n_ev[i]) for i in tf.range(n)], axis=0)
    segm=indices                                                                                                                                                                                                                        
    points = tf.gather(hypercubes, indices)                                                                                                                                                                                                                                                                                  
    xn = tf.transpose((points + tf.transpose(rnds)) * FBINS / float_me(n_strat))                                                                                                                                                                                                                                             
    n_evs = float_me(tf.gather(n_ev,indices))        
```

btw, if you want to use int64 you can use the environmental variable `VEGASFLOW_INT=64` and then DTYPEINT becomes int64 and `int_me` casts to int64.

Reply:
Ok , so `indices` is still computed in `generating_samples_in_hypercubes`, I thought otherwise.
Thanks for the feedback!

Reply:
Yes, in my first message I didn't think about the fact that the shape of indices changes from iteration to iteration so passing it around can become more of a hassle than anything else.
The only change with respect your version is manually using the `tf.repeat` element by element instead of letting `tensorflow` do the whole thing at once, because when it does it breaks (even though it should be doing the same thing as I am!)

Reply:
This is the benchmark using a gaussian in dim 8 and 12.
On the right there are the iterations needed to reach a certain required percent uncertainty, on the left the percent uncertainty reached (after at max 100 iterations if it does not reach the percent uncertainty required ).
Legend:
- vf: standard vegasflow
- vf+ not adaptive: VegasFlowPlus without redistribution of samples
- vf+ max samples per hcube 1e3: VegasFlowPlus with a limit of 1e3 samples per hypercube
- vf+ max samples per hcube 1e4: VegasFlowPlus with a limit of 1e4 samples per hypercube 
- vf+ with numpy indices: VegasFlowPlus with the indices computed using np and not tf
![gauss8d](https://user-images.githubusercontent.com/49183315/114679493-7bf8bf80-9d0c-11eb-8e49-2b1accc11225.png)
![gauss12d](https://user-images.githubusercontent.com/49183315/114679512-7f8c4680-9d0c-11eb-9b39-4328bbcae400.png)

It seems that the percent uncertainty reached is quite similar for all the possible implementation of vf+ (with a limit on the samples per hcube or using numpy indices).
Should I do some other benchmarks (different integrand or gaussian in higher dimensions)?

Reply:
This is more or less what I expected. I think this means for larger number of dimensions there are three options which are more or less equivalent:
1. Not allow to use adaptative at all (good because is the simplest one, bad because there might be integrands for which it happens to be useful).
2. Set a (small) number of samples per hypercube (I'm not sure I like this solution because we might bias the result without he user knowing).
3. Setting a maximum number of hypercubes per dimension. Right now this is enforced [here](https://github.com/N3PDF/vegasflow/blob/1426f37c7303678a057db448daf0c243af03a4a7/src/vegasflow/vflowplus.py#L109) but it can grow a lot. For instance, 8 dimensions `1e6` gives you just 4 hcubes per dim, which is 65536 in total. Changing that to 2 gives you just 256. If there's any advantage to gain from the adaptation (see point 1) you might already get it.

I favor option 3. It can be done naively, for instance if we want to have at max say 10000 hcubes we can just say that `if pow(nstrat, dim) > 1e4: nstrat = floor(pow(1e4, 1/ndim))` and it can be a variable that the user can control. That way we can set it so that by default when it goes past 13 dimensions you effectively turned `adaptative=off` but if the user is adamant on using it can change it at pleasure.

What do you think?

Reply:
@andreapasquale97 thanks for this. I agree with option 3, it is a good compromise. Let me suggest to implement and include option 3 in these benchmark plots and then rerun the full benchmarks (including hep integrands).

Reply:
From what I observed only in the first couples of iterations there may be an hypercube which has more than `1e3` or `1e4` samples for `1e6` samples, after those ones the maximum number of samples per hypercubes settles below those limits. In many cases is the first redistribution which can generate a very large number of samples in one particular hypercube. 
Because of this I thought that option 2 was better.

I haven't tried option 3 yet but I have two concerns:
- by limiting the number of hypercubes we also loose in accuracy since the error goes as (n_hypercubes)^(-1/2) and no longer as (n_events)^(-1/2) like in vegasflow
- I'm worried that even if we put a limit in the number of hypercubes if the first redistribution generates too many samples in one hypercubes there may be some problems.

I can update the previous benchmark including option 3 to see if there are any problems.
If it works fine I agree that option 3 is the better solution.

Reply:
> by limiting the number of hypercubes we also loose in accuracy since the error goes as (n_hypercubes)^(-1/2) and no longer as (n_events)^(-1/2) like in vegasflow

For your plots it seems that for a large number of dimensions there is no difference. And interesting idea (that Lepage proposes) is to have a variable number of hypercubes per dimension for larger dimension: https://arxiv.org/pdf/2009.05112.pdf

This would be a very different approach where, after the first iteration you check which dimensions seemed to make use of stratified sampling the most and reduce the number of hypercubes in the other dimensions. I guess this is the most elegant but at the same time most complicated approach (how to define "which dimensions benefit the most" is not clear to me). Maybe good for a Vegas++ :P

> I'm worried that even if we put a limit in the number of hypercubes if the first redistribution generates too many samples in one hypercubes there may be some problems.

I think the problems were a combination of the number of hypI'm worried that even if we put a limit in the number of hypercubes if the first redistribution generates too many samples in one hypercubes there may be some problems.ercubes and the number of events per hypercube.